server/src/test/java/org/elasticsearch/cluster/coordination/NodeJoinTests.java

```diff
@@ -0,0 +1,399 @@
+        when(transportService.getLocalNode()).thenReturn(initialState.nodes().getLocalNode());
+        coordinator = new Coordinator(Settings.EMPTY,
+            transportService,
+            ESAllocationTestCase.createAllocationService(Settings.EMPTY),
+            masterService,
+            () -> new CoordinationStateTests.InMemoryPersistedState(term, initialState));
+        coordinator.start();
+        coordinator.startInitialJoin();
+    }
+
+    protected DiscoveryNode newNode(int i) {
+        return newNode(i, randomBoolean());
+    }
+
+    protected DiscoveryNode newNode(int i, boolean master) {
+        Set<DiscoveryNode.Role> roles = new HashSet<>();
+        if (master) {
+            roles.add(DiscoveryNode.Role.MASTER);
+        }
+        final String prefix = master ? "master_" : "data_";
+        return new DiscoveryNode(prefix + i, i + "", buildNewFakeTransportAddress(), emptyMap(), roles, Version.CURRENT);
+    }
+
+    static class SimpleFuture extends BaseFuture<Void> {
+        final String description;
+
+        SimpleFuture(String description) {
+            this.description = description;
+        }
+
+        public void markAsDone() {
+            set(null);
+        }
+
+        public void markAsFailed(Throwable t) {
+            setException(t);
+        }
+
+        @Override
+        public String toString() {
+            return "future [" + description + "]";
+        }
+    }
+
+    private SimpleFuture joinNodeAsync(final JoinRequest joinRequest) {
+        final SimpleFuture future = new SimpleFuture("join of " + joinRequest + "]");
+        logger.debug("starting {}", future);
+        // clone the node before submitting to simulate an incoming join, which is guaranteed to have a new
+        // disco node object serialized off the network
+        coordinator.handleJoinRequest(joinRequest, new JoinHelper.JoinCallback() {
+            @Override
+            public void onSuccess() {
+                logger.debug("{} completed", future);
+                future.markAsDone();
+            }
+
+            @Override
+            public void onFailure(Exception e) {
+                logger.error(() -> new ParameterizedMessage("unexpected error for {}", future), e);
+                future.markAsFailed(e);
+            }
+        });
+        return future;
+    }
+
+    private void joinNode(final JoinRequest joinRequest) {
+        FutureUtils.get(joinNodeAsync(joinRequest));
+    }
+
+    private void joinNodeAndRun(final JoinRequest joinRequest) {
+        SimpleFuture fut = joinNodeAsync(joinRequest);
+        deterministicTaskQueue.runAllTasks(random());
+        assertTrue(fut.isDone());
+        FutureUtils.get(fut);
+    }
+
+    public void testJoinWithHigherTermElectsLeader() {
+        DiscoveryNode node0 = newNode(0, true);
+        DiscoveryNode node1 = newNode(1, true);
+        setupFakeMasterServiceAndCoordinator(1, initialState(false, node0, 1, 1,
```

Comment:
Throughout the tests, could we pick more variable terms and versions? All these `1`s and `2`s are tricky to keep track of.

Reply:
fixed in 7dcd6561201